### PR TITLE
chore: add peer dependencies for yarn pnp

### PR DESCRIPTION
**Pull request by aicest**
**What kind of change does this PR introduce?**

bugfix: fix `peerDependencies` for Yarn PnP.

**Did you add tests for your changes?**

no.

**If relevant, did you update the documentation?**

no.

**Summary**

Support Yarn PnP.

1. `@webpack-cli/serve` require `webpack-dev-server`(a peer dependency);
2. `webpack-cli` require `@webpack-cli/serve`;
3. so, `webpack-cli` would add `webpack-dev-server` into `peerDependencies`.

**Does this PR introduce a breaking change?**

no.

**Other information**

Repo for reproducing bug: https://github.com/aicest/webpack-v5-demo

image

### CODE REVIEW

File: packages/webpack-cli/package.json [hunk 1]
```diff
@@ -43,7 +43,8 @@
     "webpack-merge": "^4.2.2"
   },
   "peerDependencies": {
-    "webpack": "4.x.x || 5.x.x"
+    "webpack": "4.x.x || 5.x.x",
+    "webpack-dev-server": "3.x.x"
```
Comment (snitin315) [inline packages/webpack-cli/package.json:47]: Maybe add it in `peerDependenciesMeta`?

**Review by evilebottnawi — CHANGES_REQUESTED** [review]
This will not fix your problem, please read again the message, you are trying to fix in `webpack-cli` package, but problems in `@webpack-cli/serve`, but you can use what it is fixed, double check your versions

**Comment by aicest** [comment]
@evilebottnawi IT WILL ([I DID TRY](https://github.com/aicest/webpack-v5-demo)).

`@webpack-cli/serve` require a peer dependency (neither `devDependencies ` nor `dependencies`),
and `webpack-cli` require `@webpack-cli/serve`,
then `webpack-cli` should add the dependency (to `devDependencies/dependencies` or `peerDependencies`).

P.S. Another solution: `@webpack-cli/serve` add `webpack-dev-server` to `devDependencies/dependencies` (not a peer dependency).

**Comment by evilebottnawi** [comment]
webpack-dev-server should be not in `dependencies`, `devDependencies` will not work for your

**Comment by evilebottnawi** [comment]
https://github.com/webpack/webpack-cli/blob/master/packages/serve/package.json#L19

**Comment by ylemkimon** [comment]
@evilebottnawi Actually, this would be the right way to solve it. See #1825 and https://yarnpkg.com/advanced/error-codes/#yn0002---missing_peer_dependency.

Dependency tree (D: dependency, P: peer dependency, PM: optional peer dependency, *: added in this PR, **: reason):
```
webpack-cli
├─P> webpack
├─PM> webpack-dev-server*
├─D> @webpack-cli/info
│    └─P> webpack-cli
├─D> @webpack-cli/init
│    ├─P> webpack
│    ├─P> webpack-cli
│    ├─D> @webpack-cli/generators
│    │    ├─P> webpack
│    │    ├─P> webpack-cli
│    │    ├─D> @webpack-cli/utils
│    │    │    ├─P> webpack
│    │    │    └─P> webpack-cli
│    │    └─D> @webpack-cli/webpack-scaffold
│    └─D> @webpack-cli/utils
│         ├─P> webpack
│         └─P> webpack-cli
└─D> @webpack-cli/serve
     ├─P> webpack-cli
     └─PM> webpack-dev-server**
```
 
This will work out of the box, since Yarn 2 falls back to root dependency and `webpack-dev-server` is usually root dependency, but in case of `pnpFallbackMode: none` and workspaces, the fix is required.

**Comment by evilebottnawi** [comment]
webpack-cli doesn't use `webpack-dev-server`, it is not fixed your problem

**Comment by ylemkimon** [comment]
> webpack-cli doesn't use `webpack-dev-server`, it is not fixed your problem

However, `@webpack-cli/serve`, a dependency of `webpack-cli`, optionally requires `webpack-dev-server`, so `webpack-dev-server` should be declared as an optional peer dependency for Yarn to properly resolve dependency tree.

**Comment by arcanis** [comment]
From what I understand of the situation, @ylemkimon is correct - peer dependencies must be referenced at every level of the dependency chain, even in the intermediary packages that don't use it themselves. Cf this [article](https://dev.to/arcanis/implicit-transitive-peer-dependencies-ed0) I wrote about this some time ago.

**Review by evilebottnawi — APPROVED** [review]

**Comment by evilebottnawi** [comment]
@ylemkimon Please accept CLA

**Comment by ylemkimon** [comment]
@aicest is the author :smile:

**Comment by evilebottnawi** [comment]
@aicest Can you accept CLA?

**Review by snitin315 — APPROVED** [review]

**Comment by aicest** [comment]
@evilebottnawi CLA accepted.

**Review by anshumanv — APPROVED** [review]

**Comment by evilebottnawi** [comment]
@aicest sorry, it was fixed in other PR, anyway big thanks for the PR

**Pull request closed without merging**